src/import/speed_limiter.rs

```diff
@@ -0,0 +1,606 @@
+    /// instant must be later than the second one.
+    fn sub(later: Self::Instant, earlier: Self::Instant) -> Duration;
+
+    /// Sleeps the current thread for the given duration.
+    fn sleep(&self, dur: Duration);
+}
+
+/// The `StandardClock` obtains time using the physical clock.
+pub struct StandardClock;
+
+impl Clock for StandardClock {
+    type Instant = Instant;
+
+    fn now(&self) -> Instant {
+        Instant::now_coarse()
+    }
+
+    fn sub(later: Instant, earlier: Instant) -> Duration {
+        later.duration_since(earlier)
+    }
+
+    fn sleep(&self, dur: Duration) {
+        sleep(dur);
+    }
+}
+
+/// The type that enforces a speed limit across multiple threads.
+pub struct SpeedLimiter<C: Clock = StandardClock> {
+    // The maximum speed limit, unit = Byte/s.
+    speed_limit: f64,
+    // The last update time and the remaining amount of bytes can be sent during
+    // this second.
+    value: Mutex<(C::Instant, f64)>,
+    // The clock used for time calculation.
+    clock: C,
+}
+
+const LOG_THRESHOLD: Duration = Duration::from_secs(1);
+
+/// `SpeedLimiter` is a type to control the upper limit speed of an action
+/// cooperatively.
+impl<C: Clock> SpeedLimiter<C> {
+    /// Creates a speed limiter.
+    pub fn new(speed_limit: f64, clock: C) -> Self {
+        Self {
+            speed_limit,
+            value: Mutex::new((clock.now(), speed_limit)),
+            clock,
+        }
+    }
+
+    /// Takes a block of the given size (in bytes) from the speed limiter. If
+    /// the block is taken too quickly, this method will sleep until the speed
+    /// limit is below the specified value.
+    ///
+    /// This method will block the entire thread if the speed limit is violated,
+    /// and thus should not be invoked when implementing an async fn.
+    pub fn take(&self, tag: &str, byte_size: u64) {
+        let size = byte_size as f64;
+
+        loop {
+            let mut lock = self.value.lock().unwrap();
+            let (last_update, mut value) = *lock;
+            let now = self.clock.now();
+            let elapsed = duration_to_sec(C::sub(now, last_update));
+            value = self.speed_limit.min(value + elapsed * self.speed_limit);
+            if value >= size || value >= self.speed_limit {
+                // The second condition allows take() to proceed on a completely
+                // filled bucket, i.e. allows sending a 1 GB file on a 512 MB/s
+                // limit. The bucket value will temporarily drop to -512 MB,
+                // which will be refilled after 2 seconds as expected.
+                *lock = (now, value - size);
+                return;
+            }
+
+            // Reaching here means the bucket's value is not high enough to send
+            // all bytes. We sleep for the time needed to completely refill the
+            // bucket.
+            drop(lock);
+
+            let sleep_seconds = 1.0 - value / self.speed_limit;
```

Comment:
The formula `(size - value) / self.speed_limit` means we wait until the bucket has just enough tokens, and send `size` immediately. When `size < self.speed_limit` this does help spreading the workload evenly instead of concentrating near the start of each second.

But when `size ≫ self.speed_limit`, it means the large file would potentially need to wait much longer before being sent. Consider this situation (speed limit = 1 MB/s):

1. t=0, send 500 KB of data ⇒ value left = 500 KB
2. t=0, send 1 GB of data, not enough token, wait (1 GB - 500 KB) / 1 MB = 999.5s
3. t=0.5, bucket has been refilled to 1 MB now
4. t=999, send 600 KB of data ⇒ value left = 400 KB
5. t=999.5, the 1 GB thread wakes up, note that the bucket's value is 900 KB and still fully filled, so wait for another (1 TB - 900 KB) / 1 MB = 999.1s again
6. t=1998, send 200 KB of data ⇒ value = 800 KB
7. t=1998.1, rinse and repeat...

OTOH, with `1 - value / speed_limit`, this means before the 1 GB file is sent it will wake up every 1s, so there's a greater chance the data can actually be sent.

Maybe we could make it `(speed_limit.min(size) - value) / speed_limit`?

Reply:
Ping @Connor1996

Reply:
Sorry for the delay, got it.